### PR TITLE
Ft/os upgrade

### PR DESCRIPTION
**Pull request by ftheirs**
Upgrade to OS version V0.13.3. This PR includes the minimal changes that are needed to execute SNOS for blocks with the latest OS version.

Changes on deserialization from OS output
Some new hints that are needed for deserialization
Disable test that compares SNOS with reference PIE ---> Need to update the reference PIE with a new one (changes in output structure)

Issue Number: N/A

## Type

- [x] feature
- [ ] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: crates/starknet-os/src/config.rs [hunk 1]
```diff
@@ -32,9 +32,9 @@ pub const COMPILED_CLASS_HASH_COMMITMENT_TREE_HEIGHT: usize = 251;
 pub const CONTRACT_STATES_COMMITMENT_TREE_HEIGHT: usize = 251;
 pub const DEFAULT_INNER_TREE_HEIGHT: u64 = 64;
 // TODO: update with relevant address
-pub const DEFAULT_FEE_TOKEN_ADDR: &str = "482bc27fc5627bf974a72b65c43aa8a0464a70aab91ad8379b56a4f17a84c3";
-pub const DEFAULT_DEPRECATED_FEE_TOKEN_ADDR: &str = "482bc27fc5627bf974a72b65c43aa8a0464a70aab91ad8379b56a4f17a84c3";
-pub const SEQUENCER_ADDR_0_13_2: &str = "0x795488c127693ffb36733cc054f9e2be39241a794a4877dc8fc1dbe52750488";
+pub const DEFAULT_FEE_TOKEN_ADDR: &str = "7ce4aa542d72a82662cda96b147da9b041ecf8c61f67ef657f3bbb852fc698f";
+pub const DEFAULT_DEPRECATED_FEE_TOKEN_ADDR: &str = "5195ba458d98a8d5a390afa87e199566e473d1124c07a3c57bf19813255ac41";
+pub const SEQUENCER_ADDR_0_13_2: &str = "0x31c641e041f8d25997985b0efe68d0c5ce89d418ca9a127ae043aebed6851c5";
```
Comment (HermanObst) [inline crates/starknet-os/src/config.rs:37]: Why this is changing? could you point to where you are getting the information?

File: crates/bin/prove_block/tests/prove_block.rs [hunk 2]
```diff
@@ -49,7 +49,7 @@ const DEFAULT_COMPILED_OS: &[u8] = include_bytes!("../../../../build/os_latest.j
 #[case::inconsistent_cairo0_class_hash_1(204936)]
 #[case::no_possible_convertion_1(155007)]
 #[case::no_possible_convertion_2(155029)]
-#[case::reference_pie_with_full_output_enabled(173404)]
+// #[case::reference_pie_with_full_output_enabled(173404)]
```
Comment (HermanObst) [inline crates/bin/prove_block/tests/prove_block.rs:52]: Dont remove this test:
The failure is related to the fact that this test is also comparing output and in this version the output structure has changed. 

Fix: Find a new v0.13.3 PIE (we need them form Starkware) and modify this test to check that Block and output.

File: crates/starknet-os/src/config.rs [hunk 3]
```diff
@@ -114,8 +112,7 @@ impl StarknetGeneralConfig {
 
     pub fn empty_block_context(&self) -> BlockContext {
         let mut versioned_constants = VersionedConstants::default();
-        versioned_constants.invoke_tx_max_n_steps = self.invoke_tx_max_n_steps;
-        versioned_constants.validate_max_n_steps = self.validate_max_n_steps;
```
Comment (HermanObst) [inline crates/starknet-os/src/config.rs:118]: Dont you need to change it for:
`pub validate_max_n_steps_override: u32` ?

File: crates/starknet-os/src/hints/output.rs [hunk 4]
```diff
@@ -144,6 +162,58 @@ pub fn set_state_updates_start(
...
+) -> Result<(), HintError> {
+    let use_kzg_da_felt = exec_scopes.get::<Felt252>(vars::scopes::USE_KZG_DA)?;
+
+    let use_kzg_da = if use_kzg_da_felt == Felt252::ONE {
+        Ok(true)
+    } else if use_kzg_da_felt == Felt252::ZERO {
+        Ok(false)
+    } else {
+        Err(HintError::CustomHint("ids.use_kzg_da is not a boolean".to_string().into_boxed_str()))
+    }?;
```
Comment (HermanObst) [inline crates/starknet-os/src/hints/output.rs:186]: Add unit test for this hint

File: crates/starknet-os/src/hints/output.rs [hunk 5]
```diff
@@ -144,6 +162,58 @@ pub fn set_state_updates_start(
...
+    _exec_scopes: &mut ExecutionScopes,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+    constants: &HashMap<String, Felt252>,
+) -> Result<(), HintError> {
+    let n_actual_updates = get_integer_from_var_name(vars::ids::N_ACTUAL_UPDATES, vm, ids_data, ap_tracking)?;
+    let n_updates_small_packing_bound = get_constant(vars::ids::N_UPDATES_SMALL_PACKING_BOUND, constants)?;
+
+    let is_n_updates_small =
+        if n_actual_updates < *n_updates_small_packing_bound { Felt252::ZERO } else { Felt252::ONE };
```
Comment (HermanObst) [inline crates/starknet-os/src/hints/output.rs:212]: This should be the other way around
```suggestion
        if n_actual_updates < *n_updates_small_packing_bound { Felt252::ONE } else { Felt252::ZERO };
```

File: crates/starknet-os/src/hints/output.rs [hunk 6]
```diff
@@ -215,4 +285,35 @@ mod tests {
...
+
+    use rstest::rstest;
+
+    #[rstest]
+    // small updates
+    #[case(10, 0)]
+    #[case(255, 0)]
+    // big updates
+    #[case(256, 1)]
+    #[case(1024, 1)]
```
Comment (HermanObst) [inline crates/starknet-os/src/hints/output.rs:297]: This conditions are also the other way around, right?

**Review by HermanObst — CHANGES_REQUESTED** [review]
Good job man, I left some changes requested

File: crates/bin/prove_block/src/rpc_utils.rs [hunk 7]
```diff
@@ -215,6 +215,7 @@ pub(crate) fn get_starknet_version(block_with_txs: &BlockWithTxs) -> blockifier:
         "0.13.1.1" => blockifier::versioned_constants::StarknetVersion::V0_13_1_1,
         "0.13.2" => blockifier::versioned_constants::StarknetVersion::V0_13_2,
         "0.13.2.1" => blockifier::versioned_constants::StarknetVersion::Latest,
```
[the remark this answers is not in the document]
Comment (ftheirs) [reply crates/bin/prove_block/src/rpc_utils.rs:217]: It was not available in the Blockifier version that we're using. I can double check this but updating Blockifier to a commit where it's defined lead to some extra changes that are meant for os v0.13.4 and are not needed

File: crates/starknet-os/src/config.rs (hunk 1 above)
Comment (ftheirs) [reply crates/starknet-os/src/config.rs:37]: Yeap! It's defined in `general_config.yaml` from cairo-lang:
https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/definitions/general_config.yml

File: crates/starknet-os/src/hints/output.rs (hunk 5 above)
Comment (ftheirs) [reply crates/starknet-os/src/hints/output.rs]: Good catch! Fixed this one and the deserialize_output that was comparing the wrong value 💪

**Review by GMKrieger — COMMENTED** [review]
I didn't find anything to comment on the code, but the CI tests are failing on the tests `test_kzg_da_cairo_0` and `test_kzg_da_cairo_1`

**Review by notlesh — APPROVED** [review]

**Comment by ftheirs** [comment]
> I didn't find anything to comment on the code, but the CI tests are failing on the tests `test_kzg_da_cairo_0` and `test_kzg_da_cairo_1`

Those tests depends on some hints that will be merge in the following PRs and that's why they are not passing!

**Review by HermanObst — APPROVED** [review]
Awesome job!!!! LGTM

**Pull request merged**